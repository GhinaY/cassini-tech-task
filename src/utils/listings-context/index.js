import { createContext, useCallback, useContext, useState } from 'react';

export const ListingsContext = createContext(null);

export const useListingsContext = () => useContext(ListingsContext);

function ListingsContextProvider({children}) {
  const [listings, setListings] = useState({});

  const setListingsFromArray = useCallback((newValuesArray) => {
    const newValuesObject = newValuesArray.reduce((acc, listing) => ({ ...acc, [listing.id]: listing}), {});
    setListings(newValuesObject)
  }, []);

  const updateSingleListing = useCallback((newListingValue) => {
    setListings({
      ...listings,
      [newListingValue.id]: newListingValue,
    })
  }, [listings]);

  return (
     <ListingsContext.Provider value={{listings, setListingsFromArray, updateSingleListing}}>
       {children}
     </ListingsContext.Provider>
   );
};

export default ListingsContextProvider;
