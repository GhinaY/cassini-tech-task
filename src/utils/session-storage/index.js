export function updateStorageItem(key, value) {
  value ? sessionStorage.setItem(key, value) : sessionStorage.removeItem(key);
};

export function getStorageItem(key) {
  return sessionStorage.getItem(key);
}