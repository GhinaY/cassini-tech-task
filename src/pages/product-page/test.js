import { act, fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useParams } from 'react-router-dom';
import { ListingsContext } from '../../utils/listings-context';
import mockListingsData from '../../utils/test-utils/mock-listings-data';
import ProductPage from './index';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
}));

jest.mock('../../components/category-filter', () => ({
  __esModule: true,
  default: () => (<div>category filter</div>)
}));

const mockSetListingsFromArray = jest.fn();

function renderComponent(props) {
  return(
    render(
      <ListingsContext.Provider value={{
        listings: mockListingsData,
        setListingsFromArray: mockSetListingsFromArray
      }}>
        <MemoryRouter>
          <ProductPage {...props}/>
        </MemoryRouter>
      </ListingsContext.Provider>
    )
  );
};

describe('ProductPage', () => {
  test('renders product details correctly when data is in context and id is provided in the path params', () => {
    useParams.mockImplementation(() => ({ id: mockListingsData[0].id }));
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders loading state when id is provided in the path params but data is not in context', () => {
    useParams.mockImplementation(() => ({ id: 12345678 }));
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders editing state with no prefills when creationMode is true', () => {
    useParams.mockImplementation(() => ({}));
    const { asFragment } = renderComponent({ createMode: true });

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders editing state with prefills when the Edit button is clicked', () => {
    useParams.mockImplementation(() => ({ id: mockListingsData[0].id }));
    const { asFragment } = renderComponent();

    act(() => {
      fireEvent.click(screen.getByText(/edit/i, { selector: 'button' }));
    })

    expect(asFragment()).toMatchSnapshot();
  });
});

// Definitely needs a lot more tests, including functional ones, 
// but in the interests of time I'll skip them for now.
