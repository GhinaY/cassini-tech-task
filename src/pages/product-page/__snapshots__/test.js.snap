// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage renders editing state with no prefills when creationMode is true 1`] = `
<DocumentFragment>
  <div
    class="ProductPageContainer"
  >
    <img
      alt="the product"
      class="ProductImage"
      src="https://fakeimg.pl/400x400?text=No+img"
    />
    <form
      class="ProductDetails"
      id="listingDetailsForm"
    >
      <div>
        category filter
      </div>
      <label
        for="title"
      >
        Product title:
      </label>
      <input
        class="ProductTitle"
        name="title"
        type="text"
        value=""
      />
      <label
        for="price"
      >
        Product price:
      </label>
      <input
        name="price"
        type="number"
        value=""
      />
      <label
        for="image"
      >
        Product price:
      </label>
      <input
        name="image"
        type="url"
        value="https://fakeimg.pl/400x400?text=No+img"
      />
      <label
        for="description"
      >
        Product description:
      </label>
      <textarea
        name="description"
      />
      <button
        class="Button"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ProductPage renders editing state with prefills when the Edit button is clicked 1`] = `
<DocumentFragment>
  <div
    class="ProductPageContainer"
  >
    <img
      alt="the product"
      class="ProductImage"
      src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
    />
    <form
      class="ProductDetails"
      id="listingDetailsForm"
    >
      <div>
        category filter
      </div>
      <label
        for="title"
      >
        Product title:
      </label>
      <input
        class="ProductTitle"
        name="title"
        type="text"
        value="Mens Casual Premium Slim Fit T-Shirts "
      />
      <label
        for="price"
      >
        Product price:
      </label>
      <input
        name="price"
        type="number"
        value="22.3"
      />
      <label
        for="image"
      >
        Product price:
      </label>
      <input
        name="image"
        type="url"
        value="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
      />
      <label
        for="description"
      >
        Product description:
      </label>
      <textarea
        name="description"
      >
        Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.
      </textarea>
      <button
        class="Button"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ProductPage renders loading state when id is provided in the path params but data is not in context 1`] = `
<DocumentFragment>
  <div
    class="ProductPageContainer"
  >
    <span>
      Loading...
    </span>
  </div>
</DocumentFragment>
`;

exports[`ProductPage renders product details correctly when data is in context and id is provided in the path params 1`] = `
<DocumentFragment>
  <div
    class="ProductPageContainer"
  >
    <img
      alt="the product Mens Casual Premium Slim Fit T-Shirts "
      class="ProductImage"
      src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
    />
    <div
      class="ProductDetails"
    >
      <a
        class="BackButton"
        href="/"
      >
        &lt; Back to listings
      </a>
      <span
        class="ProductCategory"
      >
        men's clothing
      </span>
      <h1
        class="ProductTitle"
      >
        Mens Casual Premium Slim Fit T-Shirts 
      </h1>
      <span>
        ID: 2
      </span>
      <span>
        Â£22.3
      </span>
      <p
        class="ProductDescription"
      >
        Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.
      </p>
      <button
        class="Button"
      >
        Edit
      </button>
    </div>
  </div>
</DocumentFragment>
`;
